src/rust/src/x509.rs

```diff
@@ -43,6 +43,502 @@ lazy_static::lazy_static! {
+                        Some(data) => parse_general_subtrees(py, data)?,
+                        None => py.None(),
+                    };
+                    let excluded_subtrees = match nc.excluded_subtrees {
+                        Some(data) => parse_general_subtrees(py, data)?,
+                        None => py.None(),
+                    };
+                    Ok(Some(
+                        x509_module
+                            .getattr("NameConstraints")?
+                            .call1((permitted_subtrees, excluded_subtrees))?,
+                    ))
+                } else {
+                    Ok(None)
+                }
+            },
+        )
+    }
+    // This getter exists for compatibility with pyOpenSSL and will be removed.
+    // DO NOT RELY ON IT. WE WILL BREAK YOU WHEN WE FEEL LIKE IT.
+    #[getter]
+    fn _x509<'p>(
+        slf: pyo3::pycell::PyRef<'_, Self>,
+        py: pyo3::Python<'p>,
+    ) -> Result<&'p pyo3::PyAny, PyAsn1Error> {
+        let cryptography_warning = py.import("cryptography.utils")?.getattr("DeprecatedIn35")?;
+        let warnings = py.import("warnings")?;
+        warnings.call_method1(
+            "warn",
+            (
+                "This version of cryptography contains a temporary pyOpenSSL fallback path. Upgrade pyOpenSSL now.",
+                cryptography_warning,
+            ),
+        )?;
+        let backend = py
+            .import("cryptography.hazmat.backends.openssl.backend")?
+            .getattr("backend")?;
+        Ok(backend.call_method1("_cert2ossl", (slf,))?)
+    }
+}
+
+fn cert_version(py: pyo3::Python<'_>, version: u8) -> Result<&pyo3::PyAny, PyAsn1Error> {
+    let x509_module = py.import("cryptography.x509")?;
+    match version {
+        0 => Ok(x509_module.getattr("Version")?.get_item("v1")?),
+        2 => Ok(x509_module.getattr("Version")?.get_item("v3")?),
+        _ => Err(PyAsn1Error::from(pyo3::PyErr::from_instance(
+            x509_module
+                .getattr("InvalidVersion")?
+                .call1((format!("{} is not a valid X509 version", version), version))?,
+        ))),
+    }
+}
+
+#[pyo3::prelude::pyfunction]
+fn load_pem_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<Certificate> {
+    let parsed = pem::parse(data)?;
+    if parsed.tag != "CERTIFICATE" {
+        return Err(PyAsn1Error::from(pyo3::exceptions::PyValueError::new_err(
+            "Valid PEM but no BEGIN CERTIFICATE/END CERTIFICATE delimiters. Are you sure this is a certificate?"
+        )));
+    }
+    load_der_x509_certificate(py, &parsed.contents)
+}
+
+#[pyo3::prelude::pyfunction]
+fn load_der_x509_certificate(py: pyo3::Python<'_>, data: &[u8]) -> PyAsn1Result<Certificate> {
+    let raw = OwnedRawCertificate::try_new(data.to_vec(), |data| asn1::parse_single(data))?;
+    // Parse cert version immediately so we can raise error on parse if it is invalid.
+    cert_version(py, raw.borrow_value().tbs_cert.version)?;
```

Comment:
Should we also check that there's no extensions if version is 1?

Reply:
I have no opinion. We did not do that check previously.

Reply:
Did OpenSSL not do that check previously either? If not, I guess I don't care.